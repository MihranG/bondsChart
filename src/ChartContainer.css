.chart-wrapper{
    align-self: center;
    display: flex;
}