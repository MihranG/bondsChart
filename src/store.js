import { configureStore, createSlice } from "@reduxjs/toolkit";
import { combineReducers } from "redux";

const initialBlogStoreState = {
  activePeriod: "week",
  isLoading: false
};

const timePeriodSlice = createSlice({
  name: "timePeriod",
  initialState: initialBlogStoreState,
  reducers: {
    setLoading(state, action) {
      state.isLoading = action.payload;
    },
    setActiveTimePeriod(state, action) {
      state.activePeriod = action.payload;
    }
  }
});

const bondsDataSlice = createSlice({
  name: "bondsData",
  initialState: { data: [] },
  reducers: {
    setData(state, action) {
      state.data = action.payload;
    },
    clearData(state) {
      state.data = [];
    }
  }
});

const viewModeSlice = createSlice({
  name: "viewMode",
  initialState: { activeMode: "price" },
  reducers: {
    setActiveMode(state, action) {
      state.activeMode = action.payload;
    }
  }
});

const rootReducer = combineReducers({
  time: timePeriodSlice.reducer,
  bondsData: bondsDataSlice.reducer,
  viewMode: viewModeSlice.reducer
});

export const store = configureStore({
  reducer: rootReducer
});

export const {
  actions: { setLoading, setActiveTimePeriod },
  reducer: timePeriodReducer
} = timePeriodSlice;

export const {
  actions: { setData, clearData },
  reducer: bondsDataReducer
} = bondsDataSlice;

export const {
  actions: { setActiveMode },
  reducer: viewModeReducer
} = viewModeSlice;
