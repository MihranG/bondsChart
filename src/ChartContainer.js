import React from "react";
import "./ChartContainer.css";
import {
  LineChart,
  YAxis,
  XAxis,
  Tooltip,
  CartesianGrid,
  Line
} from "recharts";
import { connect } from "react-redux";

import { fetchData } from "./thunks";

export class ChartContainerDisconnected extends React.Component {
  componentDidMount() {
    const { data, fetchData, activePeriod } = this.props;
    if (data.length === 0) {
      fetchData(activePeriod)();
    }
  }

  render() {
    const { data, isLoading, activeViewMode } = this.props;
    return (
      <div className="chart-wrapper">
        {isLoading ? (
          <p id="loading_text" test-id="loading">
            Loading...
          </p>
        ) : (
          <LineChart
            width={400}
            height={400}
            data={data}
            margin={{ top: 5, right: 20, left: 10, bottom: 5 }}
          >
            <XAxis dataKey="name" />
            <YAxis />
            <Tooltip />
            <CartesianGrid stroke="#f5f5f5" />
            <Line
              type="monotone"
              dataKey={activeViewMode}
              stroke="#ff7300"
              yAxisId={0}
            />
          </LineChart>
        )}
      </div>
    );
  }
}

function mapStateToProps(state) {
  return {
    data: state.bondsData.data,
    isLoading: state.time.isLoading,
    activePeriod: state.time.activePeriod,
    activeViewMode: state.viewMode.activeMode
  };
}
function mapDispatchToProps(dispatch) {
  return {
    fetchData: period => dispatch(fetchData(period))
  };
}

const ChartContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(ChartContainerDisconnected);

export default ChartContainer;
