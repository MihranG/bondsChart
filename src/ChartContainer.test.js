import React from "react";
import { shallow, mount } from "enzyme";
import ChartContainer, { ChartContainerDisconnected } from "./ChartContainer";
import { Provider } from "react-redux";
import { store } from "./store";
import { LineChart } from "recharts";

describe("ChartContainer component", () => {
  let wrapper;
  beforeAll(() => {
    wrapper = mount(
      <Provider store={store}>
        <ChartContainer />
      </Provider>
    );
  });

  test("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  afterAll(() => {
    wrapper.detach();
  });
});

describe("ChartContainer Component", () => {
  let wrapper;
  let fetchFn;
  beforeAll(() => {
    const innerFn = jest.fn();
    fetchFn = jest.fn(() => innerFn);
    wrapper = shallow(
      <ChartContainerDisconnected
        isLoading={true}
        data={[]}
        fetchData={fetchFn}
      />
    );
  });

  test("should have appropriate props", () => {
    expect(wrapper.instance().props.isLoading).toBe(true);
    expect(wrapper.instance().props.data).toEqual([]);
  });

  test("should show loading when isLoading is true", () => {
    expect(
      wrapper.contains(
        <p id="loading_text" test-id="loading">
          Loading...
        </p>
      )
    ).toBe(true);
    expect(wrapper.find(LineChart).get(0)).toBe(undefined);
  });

  test("should not show loading when isLoading is false", () => {
    wrapper.setProps({ isLoading: false });
    expect(
      wrapper.contains(
        <p id="loading_text" test-id="loading">
          Loading...
        </p>
      )
    ).toBe(false);
    expect(wrapper.find(LineChart).get(0)).not.toBe(undefined);
  });

  test("should call fetchItems method when no data is provided", () => {
    expect(fetchFn.mock.calls.length).toBe(1);
  });

  afterAll(() => {
    wrapper.detach();
  });
});
