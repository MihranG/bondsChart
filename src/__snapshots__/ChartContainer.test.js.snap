// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartContainer component should match snapshot 1`] = `ReactWrapper {}`;
